<!DOCTYPE html>
<html lang="en">
<head>
        <title>charlesreid1</title>
        <meta charset="utf-8" />

        <!--
        CSS styles
        -->
        <link href="https://charlesreid1.github.io/theme/css/bootstrap.css"       rel="stylesheet" type="text/css">
        <link href="https://charlesreid1.github.io/theme/css/slate.css"              rel="stylesheet" type="text/css">
        <link href="https://charlesreid1.github.io/theme/css/font-awesome.css"       rel="stylesheet" type="text/css"/>
        <link href="https://charlesreid1.github.io/theme/css/pygment-solarized.css"  rel="stylesheet" type="text/css"/>

        <!--
        my CSS styles
        -->
        <link href="https://charlesreid1.github.io/theme/css/main.css"            rel="stylesheet" type="text/css">
        <link href="https://charlesreid1.github.io/theme/css/dox.css"             rel="stylesheet" type="text/css">


        <!--
        include Angular first
        -->
        <script type="text/javascript" src="https://charlesreid1.github.io/theme/js/angular-1.3.15.js"></script>
        <script type="text/javascript" src="https://charlesreid1.github.io/theme/js/skrollr-0.6.29.js"></script>
        <script type="text/javascript" src="https://charlesreid1.github.io/theme/js/d3-3.5.5.js"></script>

        <!--
        this seems like a bad idea.
        but if I don't put jquery first, I get errors about unrecognized $s.
        -->
        <script type="text/javascript" src="https://charlesreid1.github.io/theme/js/jquery-1.11.2.js"></script>

</head>
<body id="index">
<nav class="navbar navbar-default">
    <div class="container-fluid">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#myNavbar">
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span> 
            </button>
            <a href="https://charlesreid1.github.io" class="navbar-brand">charlesreid1.github.io</a>
        </div>
        <div>
            <div class="collapse navbar-collapse" id="myNavbar">
                <ul class="nav navbar-nav">

                    <li>
                        <a href="https://charlesreid1.com/wiki">Wiki</a>
                    </li>
                    <li>
                        <a href="https://git.charlesreid1.com/explore">Git</a>
                    </li>
                    <li>
                        <a href="http://spotify.charlesreid1.com">Music</a>
                    </li>
                    <li>
                        <a href="https://charlesreid1.com/life">Life</a>
                    </li>
                    <li>
                        <a href="https://charlesreid1.github.io">Blog</a>
                    </li>
                    <li>
                        <a href="https://charlesreid1.com/about">About Me</a>
                    </li>

                </ul>
            </div>
        </div>
    </div>
</nav>



<div class="container">

    <div class="row">
        <div class="col-md-12 col-centered" style="text-align: center;">
            <div class="v50"></div>
            <div class="cosmojumbo">
                <h1 class="jumbojumbo">charlesreid1.com blog<h1>
            </div>
        </div>
    </div>


    <div class="v50"></div>


    <div class="row">

        <div class="col-md-8" style="text-align: center;">

  	            	<article>
<header>
<div class="row">


    <div class="col-md-12">
        <div class="headtext" style="text-align: center;">
            <h1><b>
                Deconvoluting Convolutional Neural Networks
            </b></h1>
        
            <p class="lead">
                Posted 
                <time datetime="2019-05-29T14:00:00-07:00" pubdate>Wednesday 05/29/2019</time>
                in 
                <a href="https://charlesreid1.github.io/category/machine-learning.html">Machine Learning</a>
                </p>
            <p>
            <span style="font-size: 14px">
                <a href="https://charlesreid1.github.io/deconvoluting-convolutional-neural-networks.html">permalink</a>
            </span>
            </p>
        </div>

    </div>

</div>
</header>


<div class="v50"></div>


<div class="row">
    <div class="col-md-12" style="text-align: left;">

        <div class="entry-content"><h1>Table of Contents</h1>
<ul>
<li><a href="#table-of-contents">Table of Contents</a></li>
<li><a href="#introduction-a-simple-cnn-example">Introduction: A Simple CNN Example</a></li>
<li><a href="#exploring-data">Exploring Data</a></li>
<li><a href="#create-cnn">Create CNN</a></li>
<li>
<p><a href="#analyzing-network-architecture-and-tensor-shapes">Analyzing Network Architecture and Tensor Shapes</a></p>
<ul>
<li>
<p><a href="#input-image-layer">Input Image Layer</a></p>
</li>
<li>
<p><a href="#first-convolution-layer">First Convolution Layer</a></p>
</li>
<li><a href="#first-convolution-layer">First Activation Layer</a></li>
<li>
<p><a href="#first-convolution-layer">First Pooling Layer</a></p>
</li>
<li>
<p><a href="#second-convolution-layer">Second Convolution Layer</a></p>
</li>
<li><a href="#second-convolution-layer">Second Activation Layer</a></li>
<li>
<p><a href="#second-convolution-layer">Second Pooling Layer</a></p>
</li>
<li>
<p><a href="#third-convolution-layer">Third Convolution Layer</a></p>
</li>
<li><a href="#third-convolution-layer">Third Activation Layer</a></li>
<li>
<p><a href="#third-convolution-layer">Third Pooling Layer</a></p>
</li>
<li>
<p><a href="#flatten-and-dense-layers">Flatten and Dense Layers</a></p>
</li>
<li><a href="#categorical-output">Categorical Output</a></li>
</ul>
</li>
<li>
<p><a href="#image-transformer">Image Transformer</a></p>
</li>
<li>
<p><a href="#next-steps">Next Steps</a></p>
</li>
</ul>
<h1>Introduction: A Simple CNN Example</h1>
<p>As part of our weekly Deep Learning for Genomics reading group
here in the <a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology (DIB Lab)</a>,
we are applying convolutional neural networks (deep learning) 
to various problems in genomics and biology.</p>
<p>For the most recent meeting, we prepared some notes on how convolutional
neural networks work. The notes are in the form of a Jupyter notebook.
This blog post summarizes some of the important conclusions from the
notebook and links to relevant sections in the notebook.</p>
<p>In the notebook covered in this blog post, we set up a 
simple convolutional neural network from an example on the
<a href="https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html">keras blog</a>.
This example is used to classify input images as being either
a cat or a dog.</p>
<p>All materials covered in this blog post are in the 
<a href="https://github.com/charlesreid1/deconvoluting-convolutions">charlesreid1/deconvoluting-convolutions</a>
repository on Github.</p>
<h1>Exploring the Data</h1>
<p><strong>TL;DR:</strong> When developing a deep learning model for a problem,
it is important to start by exploring the data and understanding
it thoroughly. </p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Image-Data">Link to "Image Data" section of notebook</a></p>
<h1>Create CNN</h1>
<p><strong>TL;DR:</strong> Our convolutional neural network consists of the following architecture:</p>
<ul>
<li>Convolutional Stage #1<ul>
<li>Convolution (3 x 3 kernel, 32 filters)</li>
<li>Activation (ReLU)</li>
<li>Max Pooling (2x2)</li>
</ul>
</li>
<li>Convolutional Stage #2<ul>
<li>Convolution (3 x 3 kernel, 32 filters)</li>
<li>Activation (ReLU)</li>
<li>Max Pooling (2x2)</li>
</ul>
</li>
<li>Convolutional Stage #3<ul>
<li>Convolution (3 x 3 kernel, 64 filters)</li>
<li>Activation (ReLU)</li>
<li>Max Pooling (2x2)</li>
</ul>
</li>
<li>Flatten</li>
<li>Dense (64 nodes)</li>
<li>Activation (ReLU)</li>
<li>Dropout (0.5)</li>
<li>Dense (1 node)</li>
<li>Activation (ReLU)</li>
</ul>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Create-Convolutional-Neural-Network">Link to "Create Convolutional Neural Network" section of notebook</a></p>
<h1>Analyzing Network Architecture and Tensor Shapes</h1>
<p><strong>TL;DR:</strong> Each step of the neural network transforms
an input tensor of a given shape into an output tensor
of a (potentially different) shape.</p>
<p>In this section of the notebook, we step through each
of the neural network's layers to explain how the size
of each layer's inputs and outputs are determined.</p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Network-Architecture/Shapes">Link to "Network Architecture/Shapes" section of notebook</a></p>
<h2>Input Image Layer</h2>
<p><strong>TL;DR:</strong> The size of the cat and dog images is 150 x 150 pixels.
Each image is a color image, so it consists of 3 channels. Therefore,
the input to the very first layer has a shape of</p>
<div class="math">$$
(\mbox{None}, w_0, h_0, c_0) = (\mbox{None}, 150, 150, 3)
$$</div>
<p>(where "None" indicates a variable-size dimension that is equal to
the number of total input images, or alternatively, the number of 
images per batch, if we are using batch learning).</p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Input-Image-Layer">Link to "Input Image Layer" section of notebook</a></p>
<h2>First Convolution Layer</h2>
<p><strong>TL;DR:</strong> A convolutional layer with a kernel size of <span class="math">\(k_1 \times k_1\)</span>
and a number of filters <span class="math">\(c_1\)</span> will transform the shape of the input image to:</p>
<div class="math">$$
(\mbox{None}, w_1, h_1, c_1) = 
(\mbox{None}, 148, 148, 32)
$$</div>
<p>where</p>
<div class="math">$$
w_1 = w_0 - k_1 + 1 \\
h_1 = h_0 - k_1 + 1
$$</div>
<p>Importantly, each of the three input channels are added together to determine
their contribution to the final convolution filters - the number of input channels
does not affect the number of output channels. </p>
<p>The total number of output channels is equal to the number of filters
in the convolution layer. </p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#First-Convolution-Layer">Link to "First Convolutional Layer" section of notebook</a></p>
<h2>First Activation Layer</h2>
<p><strong>TL;DR:</strong> The activation layer is a straightforward one-to-one mapping -
each individual value from the output of the convolution layer is fed through
the rectified linear unit (ReLU) function and the resulting output value becomes
the input to the next layer. The ReLU function is given by:</p>
<div class="math">$$
\mbox{ReLU}(x) = \max(0,x)
$$</div>
<p>The activation layer does not change the shape of the input tensor.</p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#First-Activation-Layer">Link to "First Activation Layer" section of notebook</a></p>
<h2>First MaxPooling Layer</h2>
<p><strong>TL;DR:</strong> The max pooling layer is a way of making the final convolutional
filters (the "feature-detectors" of the convolutional neural network) less 
sensitive to the exact placement of features. The pooling layer only affects
the size of the filter, not the number of channels.</p>
<p>If we use a max pooling window of <span class="math">\(p_1 \times p_1\)</span>, we will reduce the image
size by <span class="math">\(\mbox{ceil}(w_1/p_1)\)</span> and <span class="math">\(\mbox{ceil}(h_1/p_1)\)</span>. This reduces the input tensor shape
to:</p>
<div class="math">$$
(\mbox{None}, \mbox{ceil}(w_1/p_1), \mbox{ceil}(h_1/p_1), c_1) = 
(\mbox{None}, 74, 74, 32)
$$</div>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#First-Max-Pooling-Layer">Link to "First Max Pooling Layer" section of notebook</a></p>
<h2>Second Convolution Layer</h2>
<p><strong>TL;DR:</strong> The second convolutional layer has a kernel size 
of <span class="math">\(k_2 \times k_2\)</span> and a number of filters <span class="math">\(c_2\)</span>, which will
transform the shape of the input image in the same way as 
described for the first convolutional layer.</p>
<p>Note that just as the number of channels (3) in each 
input to the first convolutional layer did not affect
the final number of channels in the output of the convolutional
layer (number of channels was fixed by specifying number of
output filters for the convolutional layer), so the number of
input channels to the second convolutional layer does not affect 
the number of output channels from the second convolutional
layer.</p>
<p>The final shape coming out of the second convolutional layer is:</p>
<div class="math">$$
(\mbox{None}, w_2, h_2, c_2) = 
(\mbox{None}, 72, 72, 32)
$$</div>
<p>where</p>
<div class="math">$$
w_2 = w_1 - k_2 + 1 \\
h_2 = h_1 - k_2 + 1 \\
$$</div>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Second-Convolution-Layer">Link to "Second Convolutional Layer" section of notebook</a></p>
<h2>Second Activation Layer</h2>
<p><strong>TL;DR:</strong> The activation layer again uses a function to
map input values to output values in a one-to-one mapping,
so the activation layer does not change the shape of the 
input tensor.</p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Second-Activation-Layer">Link to "Second Activation Layer" section of notebook</a></p>
<h2>Second MaxPooling Layer</h2>
<p><strong>TL;DR:</strong> The second max pooling layer uses a pooling
window of size <span class="math">\(p_2 \times p_2\)</span>. This will reduce the input
size to <span class="math">\(\mbox{ceil}(w_2/p_2) \times \mbox{ceil}(h_2/p_2)\)</span>. This reduces 
the input tensor shape to:</p>
<div class="math">$$
(\mbox{None}, \mbox{ceil}(w_2/p), \mbox{ceil}(h_2/p), c_2) = 
(\mbox{None}, 36, 36, 32)
$$</div>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Second-Max-Pooling-Layer">Link to "Second Max Pooling Layer" section of notebook</a></p>
<h2>Third Convolution Layer</h2>
<p><strong>TL;DR:</strong> The third convolution layer with a kernel size 
of <span class="math">\(k_3 \times k_3\)</span> and <span class="math">\(c_3\)</span> output filters will transform
the input tensor shape in the following way (note that the
third convolutional layer has 64 filters, not 32):</p>
<div class="math">$$
(\mbox{None}, w_3, h_3, c_3) =
(\mbox{None}, 34, 34, 64)
$$</div>
<p>where</p>
<div class="math">$$
w_3 = w_2 - k_3 + 1 \\
h_3 = h_2 - k_3 + 1
$$</div>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Third-Convolution-Layer">Link to "Third Convolutional Layer" section of notebook</a></p>
<h2>Third Activation Layer</h2>
<p><strong>TL;DR:</strong> The activation layer again uses a function to
map input values to output values in a one-to-one mapping,
so the activation layer does not change the shape of the 
input tensor.</p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Third-Activation-Layer">Link to "Third Activation Layer" section of notebook</a></p>
<h2>Third MaxPooling Layer</h2>
<p><strong>TL;DR:</strong> The thid max pooling layer uses a pooling
window of size <span class="math">\(p_3 \times p_3\)</span>. This will reduce the input
size to <span class="math">\(\mbox{ceil}(w_3/p_3) \times \mbox{ceil}(h_3/p_3)\)</span>. This reduces 
the input tensor shape to:</p>
<div class="math">$$
(\mbox{None}, \mbox{ceil}(w_3/p_3), \mbox{ceil}(h_3/p_3), c_2) = 
(\mbox{None}, 17, 17, 64)
$$</div>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Third-Max-Pooling-Layer">Link to "Third Max Pooling Layer" section of notebook</a></p>
<h2>Flatten and Dense Layers</h2>
<p><strong>TL;DR:</strong> The flatten layer converts a tensor of dimension <span class="math">\((\mbox{None}, 17, 17, 64)\)</span>
into a 1D vector of <span class="math">\(17 \times 17 \times 64 = 18,496\)</span> neural network nodes. This does not
change any of the values, it simply reshapes the input tensor.</p>
<p>The first dense layer reduces the flattened <span class="math">\(18,496\)</span> nodes to <span class="math">\(64\)</span> nodes, using a fully connected
layer of nodes. These values are then passed through an activation function (as with the above
activation layers, this is a one-to-one mapping and does not change the shape of the input tensor).
The dense layer is followed by a dropout layer to help prevent overfitting; this pattern is common
in convolutional neural networks.</p>
<p>The second dense layer further reduces the <span class="math">\(64\)</span> nodes to a single node, whose output will determine
whether the input image is a cat or a dog.</p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Flatten-Layer">Link to "Flatten Layer" section of notebook</a></p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Dense-(64)">Link to "Dense (64) Layers" section of notebook</a></p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Dense-(1)">Link to "Dense (1) Layers" section of notebook</a></p>
<h2>Categorical Output</h2>
<p><strong>TL;DR:</strong> Normally when classifying cats and dogs, we would have two output neurons, one to
output a binary yes/no to answer "is this a cat?" and another output a binary yes/no to answer 
"is this a dog?". However, in this example, we assume that <em>all</em> inputs contain either only cats
or only dogs, so the single-output binary classifier is determining whether an image is a dog (0)
or a cat (1).</p>
<h1>Image Transformer</h1>
<p><strong>TL;DR:</strong> The <code>ImageDataGenerator</code> class is a class provided by keras
for loading image data from a directory and (optionally) applying various
transformations to the images in order to generate additional training data
from a set of images. For example, the following code block from the 
notebook creates an <code>ImageDataGenerator</code> class that will load images from a
folder on disk, and applies various transformations (shearing, zooming, 
and horizontally flipping) to each image during the training process.</p>
<div class="highlight"><pre><span></span><span class="n">train_datagen</span> <span class="o">=</span> <span class="n">ImageDataGenerator</span><span class="p">(</span>
    <span class="n">rescale</span><span class="o">=</span><span class="mf">1.</span> <span class="o">/</span> <span class="mi">255</span><span class="p">,</span>
    <span class="n">shear_range</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
    <span class="n">zoom_range</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
    <span class="n">horizontal_flip</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>


<p>This can then be used to generate test image data:</p>
<div class="highlight"><pre><span></span>train_generator = train_datagen.flow_from_directory(
    &#39;train&#39;,
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode=&#39;binary&#39;)
</pre></div>


<p>This will look for images in the relative path <code>train/data/</code>
(note the implicit <code>data/</code> directory tacked on the end).
Note that this image data generator allows us to use images
that do not have size <span class="math">\(150 \times 150\)</span>, as they will be re-sized
to <code>target_size</code>.</p>
<p><a href="https://tinyurl.com/deconvoluting-convolutions#Image-Transformer">Link to "Image Transformer" section of notebook</a></p>
<h2>Next Steps</h2>
<p>Now that we have walked through a sample convolutional neural network
and covered how each layer transforms the size of the input tensor, 
we are ready to start applying convolutional neural networks to real
problems.</p>
<p>Our next blog post will cover the materials in the
<a href="https://github.com/charlesreid1/deep-learning-genomics">charlesreid1/deep-learning-genomics</a>
repository on Github, which applies the convolutional neural 
network concept in a 1D context (applying convolutions to 1D sequences, 
instead of 2D images) to learn about (and predict) DNA transcription factor 
binding sites.  </p>
<script type="text/javascript">if (!document.getElementById('mathjaxscript_pelican_#%@#$@#')) {
    var align = "center",
        indent = "0em",
        linebreak = "false";

    if (false) {
        align = (screen.width < 768) ? "left" : align;
        indent = (screen.width < 768) ? "0em" : indent;
        linebreak = (screen.width < 768) ? 'true' : linebreak;
    }

    var mathjaxscript = document.createElement('script');
    mathjaxscript.id = 'mathjaxscript_pelican_#%@#$@#';
    mathjaxscript.type = 'text/javascript';
    mathjaxscript.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.3/latest.js?config=TeX-AMS-MML_HTMLorMML';
    mathjaxscript[(window.opera ? "innerHTML" : "text")] =
        "MathJax.Hub.Config({" +
        "    config: ['MMLorHTML.js']," +
        "    TeX: { extensions: ['AMSmath.js','AMSsymbols.js','noErrors.js','noUndefined.js'], equationNumbers: { autoNumber: 'AMS' } }," +
        "    jax: ['input/TeX','input/MathML','output/HTML-CSS']," +
        "    extensions: ['tex2jax.js','mml2jax.js','MathMenu.js','MathZoom.js']," +
        "    displayAlign: '"+ align +"'," +
        "    displayIndent: '"+ indent +"'," +
        "    showMathMenu: true," +
        "    messageStyle: 'normal'," +
        "    tex2jax: { " +
        "        inlineMath: [ ['\\\\(','\\\\)'] ], " +
        "        displayMath: [ ['$$','$$'] ]," +
        "        processEscapes: true," +
        "        preview: 'TeX'," +
        "    }, " +
        "    'HTML-CSS': { " +
        "        styles: { '.MathJax_Display, .MathJax .mo, .MathJax .mi, .MathJax .mn': {color: 'inherit ! important'} }," +
        "        linebreaks: { automatic: "+ linebreak +", width: '90% container' }," +
        "    }, " +
        "}); " +
        "if ('default' !== 'default') {" +
            "MathJax.Hub.Register.StartupHook('HTML-CSS Jax Ready',function () {" +
                "var VARIANT = MathJax.OutputJax['HTML-CSS'].FONTDATA.VARIANT;" +
                "VARIANT['normal'].fonts.unshift('MathJax_default');" +
                "VARIANT['bold'].fonts.unshift('MathJax_default-bold');" +
                "VARIANT['italic'].fonts.unshift('MathJax_default-italic');" +
                "VARIANT['-tex-mathit'].fonts.unshift('MathJax_default-italic');" +
            "});" +
            "MathJax.Hub.Register.StartupHook('SVG Jax Ready',function () {" +
                "var VARIANT = MathJax.OutputJax.SVG.FONTDATA.VARIANT;" +
                "VARIANT['normal'].fonts.unshift('MathJax_default');" +
                "VARIANT['bold'].fonts.unshift('MathJax_default-bold');" +
                "VARIANT['italic'].fonts.unshift('MathJax_default-italic');" +
                "VARIANT['-tex-mathit'].fonts.unshift('MathJax_default-italic');" +
            "});" +
        "}";
    (document.body || document.getElementsByTagName('head')[0]).appendChild(mathjaxscript);
}
</script></div>

    </div>
</div>

<div class="v20"></div>

<div class="row">
    <div class="col-md-12" style="text-align: left;">

        <div class="tags">
            <p>Tags:&nbsp;&nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/deep-learning.html">deep learning</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/machine-learning.html">machine learning</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/neural-networks.html">neural networks</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/python.html">python</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/keras.html">keras</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/convolutional-neural-networks.html">convolutional neural networks</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/cnn.html">cnn</a>
                    &nbsp;&nbsp;
        </div>

    </div>
</div>


<div class="v200"></div>
  	            	</article>
  	            	<article>
<header>
<div class="row">


    <div class="col-md-12">
        <div class="headtext" style="text-align: center;">
            <h1><b>
                Graphs for Bioinformatics, Part 2: Finding Eulerian Paths
            </b></h1>
        
            <p class="lead">
                Posted 
                <time datetime="2019-05-07T16:00:00-07:00" pubdate>Tuesday 05/07/2019</time>
                in 
                <a href="https://charlesreid1.github.io/category/computational-biology.html">Computational Biology</a>
                </p>
            <p>
            <span style="font-size: 14px">
                <a href="https://charlesreid1.github.io/graphs-for-bioinformatics-part-2-finding-eulerian-paths.html">permalink</a>
            </span>
            </p>
        </div>

    </div>

</div>
</header>


<div class="v50"></div>


<div class="row">
    <div class="col-md-12" style="text-align: left;">

        <div class="entry-content"><h1>Table of Contents</h1>
<ul>
<li><a href="#the-context-de-bruijn-graphs">The Context: de Bruijn Graphs</a></li>
<li><a href="#assembling-the-de-bruijn-graph">Assembling the de Bruijn Graph</a><ul>
<li><a href="#directed-graph-representation-adjacency-list">Directed Graph Representation: Adjacency List</a></li>
<li><a href="#a-quick-example">A Quick Example</a></li>
<li><a href="#python-vs-go">Python vs. Go</a></li>
<li><a href="#directed-graph-class-python-implementation">Directed Graph Class: Python Implementation</a></li>
</ul>
</li>
<li><a href="#checking-for-eulerian-paths-and-cycles">Checking for Eulerian Paths and Cycles</a></li>
<li><a href="#algorithm-building-blocks">Algorithm Building Blocks</a><ul>
<li><a href="#depth-first-search">Depth First Search</a></li>
<li><a href="#kosarajus-algorithm-connected-components">Kosaraju's Algorithm: Connected Components</a></li>
</ul>
</li>
<li><a href="#finding-the-eulerian-path-or-cycle">Finding the Eulerian Path or Cycle</a><ul>
<li><a href="#hierholzers-algorithm-finding-euler-cycles">Hierholzer's Algorithm: Finding Euler Cycles</a></li>
<li><a href="#hierholzers-algorithm-finding-euler-paths">Hierholzer's Algorithm: Finding Euler Paths</a></li>
</ul>
</li>
<li><a href="#final-code">Final Code</a></li>
</ul>
<p><br />
<br /></p>
<h1>The Context: de Bruijn Graphs</h1>
<p>In <a href="https://charlesreid1.github.io/graphs-for-bioinformatics-part-1-de-bruijn-graphs-hamiltonian-paths-and-eulerian-paths.html">Part 1 of this post</a>
we discussed a data structure called a de Bruijn graph and covered its application
to genome assembly. To summarize, a de Bruijn graph is a type of graph that represents
a set of k-mers as a set of directed edges on a graph, connecting the k-mer's (k-1)-mer prefix
(the source vertex) to the k-mer's (k-1)-mer suffix (the destination vertex).</p>
<p>As an example, if <span class="math">\(k = 5\)</span>, we can represent the k-mer "AAGCT" as an edge connecting the vertex
<code>AAGC</code> to the vertex <code>AGCT</code>.</p>
<p>The de Bruijn graph is used to solve a set of problems on <a href="https://rosalind.info">Rosalind.info</a>,
a website with bioinformatics programming challenges, as part of working through the
textbook <a href="http://bioinformaticsalgorithms.com/">Bioinformatics Algorithms: An Active Learning Approach</a>
and its associated website (<a href="https://rosalind.info">Rosalind.info</a>).</p>
<h1>Assembling the de Bruijn Graph</h1>
<p>The problems from Rosalind.info that require the use of a de Bruijn graph come from
Chapter 3. These problems generally give the user either a list of k-mers (to assemble
into a de Bruijn graph, as in problem <a href="http://rosalind.info/problems/ba3e/">BA3E</a>) 
or a long sequence of DNA (which can be turned into a list of
k-mers and assembled into a de Bruijn graph, as in problem <a href="http://rosalind.info/problems/ba3d/">BA3D</a>).</p>
<p>If we are starting with a long string of DNA, we can run through the entire string
and extract k-mers using a sliding window. For a string of DNA of length <span class="math">\(d\)</span>, this procedure
will create <span class="math">\(d - k + 1\)</span> k-mers.</p>
<h2>Directed Graph Representation: Adjacency List</h2>
<p>The de Bruijn graph is a directed graph. To represent this
graph in memory, we utilize an adjacency list data structure.
An adjacency list is a key-value lookup table (implemented using
a hash map) wherein each <em>source vertex</em> in the graph is a key 
in the lookup table, and the corresponding value is a list of
all <em>destination vertices</em> (all vertices that have a directed
edge starting from the source vertex and ending at that vertex).</p>
<p>A Python dictionary can be used to implement the adjacency list
hash table. The dictionary keys are the source vertices (or rather,
their string labels), and the dictionary values are a list of 
destination vertices (a list of their string labels).</p>
<p>Thus, the graph <code>AA -&gt; BB -&gt; CC -&gt; DD</code> would be represented with
the hash table:</p>
<div class="highlight"><pre><span></span>adjacency_list[&#39;AA&#39;] = [&#39;BB&#39;]
adjacency_list[&#39;BB&#39;] = [&#39;CC&#39;]
adjacency_list[&#39;CC&#39;] = [&#39;DD&#39;]
</pre></div>


<p>(Notice from this example that the keys of the adjacency list
gives a list of <em>source vertices only</em>, to get all vertices we
need to look at the values of the adjacency list too.)</p>
<h2>A Quick Example</h2>
<p>As a simple example, consider the de Bruijn graph formed from
the DNA string <code>AAGATTCTCTAC</code> and <span class="math">\(k = 4\)</span>.</p>
<p>This is first turned into a bag of <span class="math">\(d - k + 1 = 9\)</span> 
4-mers (our edges):</p>
<div class="highlight"><pre><span></span><span class="n">Sequence</span><span class="o">:</span>   <span class="n">AAGATTCTCTAC</span>
<span class="mi">4</span><span class="o">-</span><span class="n">mers</span><span class="o">:</span>     <span class="n">AAGA</span>
             <span class="n">AGAT</span>
              <span class="n">GATT</span>
               <span class="n">ATTC</span>
                <span class="n">TTCT</span>
                 <span class="n">TCTC</span>
                  <span class="n">CTCT</span>
                   <span class="n">TCTA</span>
                    <span class="n">CTAC</span>
</pre></div>


<p>Next, we also create a bag of <span class="math">\(d - k + 1 = 10\)</span> 
3-mers (vertices):</p>
<div class="highlight"><pre><span></span><span class="n">Sequence</span><span class="o">:</span>   <span class="n">AAGATTCTCTAC</span>
<span class="mi">3</span><span class="o">-</span><span class="n">mers</span><span class="o">:</span>     <span class="n">AAG</span>
             <span class="n">AGA</span>
              <span class="n">GAT</span>
               <span class="n">ATT</span>
                <span class="n">TTC</span>
                 <span class="n">TCT</span>
                  <span class="n">CTC</span>
                   <span class="n">TCT</span>
                    <span class="n">CTA</span>
                     <span class="n">TAC</span>
</pre></div>


<p>Now we can iterate over every 4-mer edge, find its
prefix 3-mer and suffix 3-mer, and create a corresponding
entry in the adjacency list hash table.</p>
<p>The list of edges looks like this:</p>
<div class="highlight"><pre><span></span>AAG -&gt; AGA
AGA -&gt; GAT
ATT -&gt; TTC
CTA -&gt; TAC
CTC -&gt; TCT
GAT -&gt; ATT
TCT -&gt; CTA,CTC
TTC -&gt; TCT
</pre></div>


<p>The corresponding dictionary should look like this:</p>
<div class="highlight"><pre><span></span>adjacency_list[&#39;AAG&#39;] = [&#39;AGA&#39;]
adjacency_list[&#39;AGA&#39;] = [&#39;GAT&#39;]
adjacency_list[&#39;ATT&#39;] = [&#39;TTC&#39;]
adjacency_list[&#39;CTA&#39;] = [&#39;TAC&#39;]
adjacency_list[&#39;CTC&#39;] = [&#39;TCT&#39;]
adjacency_list[&#39;GAT&#39;] = [&#39;ATT&#39;]
adjacency_list[&#39;TCT&#39;] = [&#39;CTA&#39;,&#39;CTC&#39;]
adjacency_list[&#39;TTC&#39;] = [&#39;TCT&#39;]
</pre></div>


<h2>Python vs Go</h2>
<p>Now that we're ready to implement a directed graph object
and populate it using the data given in the problem, we 
have to make the difficult choice of what language we
want to use to implement the directed graph.</p>
<p>We have covered our use of the Go programming language 
for Rosalind.info problems before (we have previously 
covered recursion for Chapter 2 problems in 
<a href="https://charlesreid1.github.io/recursive-backtracking-in-go-for-bioinformatics-applications-1-counting-variations.html">Part 1</a>,
<a href="https://charlesreid1.github.io/recursive-backtracking-in-go-for-bioinformatics-applications-2-generating-variations.html">Part 2</a>,
and <a href="https://charlesreid1.github.io/recursive-backtracking-in-go-for-bioinformatics-applications-3-go-implementation-of-backtracking.html">Part 3</a>
of another post, and we also wrote <a href="https://charlesreid1.github.io/learning-bioinformatics-with-go-and-rosalind.html">this post</a>
on our impression of Go and its usefulness in
bioinformatics.</p>
<p>We are also implementing all of the solutions to the
Rosalind.info problems in our Go libarary, <a href="https://github.com/charlesreid1/go-rosalind">go-rosalind</a>
(see <a href="https://godoc.org/github.com/charlesreid1/go-rosalind/rosalind">corresponding documentation on godoc.org</a>).</p>
<p><strong>However,</strong> we have learned the hard way that Go requires
a lot of boilerplate code (boilerplate code that is necessary,
mind you, because all of that boilerplate will eventually
morph into something problem-specific).</p>
<p>This all means that Go is a very cumbersome language to use
to get an algorithm prototype up and running.</p>
<p>Python, on the other hand, is a very easy language for
prototyping and has plenty of handy built-in functions
and modules that make prototyping an algorithm far 
easier and faster than doing it in Go.</p>
<p>Our strategy, therefore, is to prototype our algorithm
and corresponding graph object in Python, get the
algorithm working and tested, then convert the code 
to Go when we are finished.</p>
<h2>Directed Graph Class: Python Implementation</h2>
<p>Note that while we could simply use the dictionary object
itself as the graph data structure, this is somewhat
inelegant, and we would like instead to define a class
to bundle related behavior and data together.</p>
<p>We implement the directed graph by defining an <code>AdjacencyGraph</code>
class. This is just a glorified wrapper around the ajacency
list dictionary, with some extra methods.</p>
<p>We start by defining the class (it inherits from <code>object</code> so
it has no parent type):</p>
<div class="highlight"><pre><span></span><span class="kr">class</span> <span class="nx">AdjacencyGraph</span><span class="p">(</span><span class="nx">object</span><span class="p">)</span><span class="o">:</span>
    <span class="s2">&quot;&quot;&quot;Directed graph stored using an adjacency list&quot;&quot;&quot;</span>
    <span class="nx">def</span> <span class="nx">__init__</span><span class="p">(</span><span class="nx">self</span><span class="p">)</span><span class="o">:</span>
        <span class="s2">&quot;&quot;&quot;Constructor&quot;&quot;&quot;</span>
        <span class="nx">self</span><span class="p">.</span><span class="nx">adj</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="nx">self</span><span class="p">.</span><span class="nx">dfs_started</span> <span class="o">=</span> <span class="nx">False</span>
</pre></div>


<p>The constructor just initializes an empty adjacency list dictionary.</p>
<p>We also define two built-in methods for convenience:
<code>__str__</code> for the string representation of the graph 
(so we can pass the graph object to <code>print()</code>), and
<code>__len__</code> for getting the number of (source) vertices 
on the graph.</p>
<div class="highlight"><pre><span></span>    def __str__(self):
        &quot;&quot;&quot;String representation&quot;&quot;&quot;
        s = []
        for source in self.adj.keys():
            sinks = self.adj[source]
            for sink in sinks:
                m = &quot;%s -&gt; %s\n&quot;%(source,sink)
                s.append(m)
        return &quot;&quot;.join(s)

    def __len__(self):
        &quot;&quot;&quot;Number of vertices on graph&quot;&quot;&quot;
        s = set()
        for source in self.adj.keys():
            s.add(source)
            for sink in self.adj[k]:
                s.add(sink)
        return len(s)
</pre></div>


<p>Next, we define some basic functionality useful for all graphs:</p>
<ul>
<li>Getting the in-degree and out-degree of a vertex</li>
</ul>
<div class="highlight"><pre><span></span>    def in_degree(self,u):
        n = 0
        for v in self.adj.keys():
            sinks = self.adj[v]
            if u in sinks:
                n += 1
        return n

    def out_degree(self,u):
        if u in self.adj.keys():
            return len(self.adj[u])
        else:
            return 0
</pre></div>


<p>We also define a generator for creating vertices:</p>
<div class="highlight"><pre><span></span>    def vertices(self):
        vertices = set()
        for k in self.adj.keys():
            vertices.add(k)
            for m in self.adj[k]:
                vertices.add(m)
        for v in vertices:
            yield v

    def n_vertices(self):
        return len(self)

    def n_edges(self):
        n = 0
        for source in self.adj.keys():
            try:
                n += len(self.adj[source])
            except:
                # in case value is None
                pass
        return n

    def get_neighbors(self,u):
        &quot;&quot;&quot;Get all neighbors of node u&quot;&quot;&quot;
        # Note: neighbors are stored in
        # sorted order
        if u in self.adj.keys():
            return self.adj[u]
        else:
            return []
</pre></div>


<p>Finally, we add a method <code>add_edge()</code> that allows us
to create an edge from vertex u to vertex v (and add
the vertices to the graph if either do not yet exist
on the graph).</p>
<p>For convenience, we maintain the adjacency list values
(the list of destination vertices) in lexicographic
order.</p>
<div class="highlight"><pre><span></span>    def add_edge(self, u, v):
        &quot;&quot;&quot;Add an edge from u to v&quot;&quot;&quot;
        # For each source vertex:
        if u in self.adj.keys():

            # Get existing sink list
            t = self.adj[u]

            # Append to it
            t.append(v)

            # Keep list of sinks sorted
            # (lexicographic string sorting)
            t.sort()

            # Create the new edge 
            # from source to sink
            self.adj[u] = t

        else:
            # Initialize the list of sinks (v)
            # for the given source (u)
            self.adj[u] = [v]
</pre></div>


<p>Now, to assemble the de Bruijn graph, we can iterate
over every k-mer edge, form the prefix and suffix vertices,
and call the <code>add_edge()</code> function on the graph.</p>
<h1>Checking for Eulerian Paths and Cycles</h1>
<p>To recap Eulerian paths versus Eulerian cycles (discussed in
<a href="https://charlesreid1.github.io/graphs-for-bioinformatics-part-1-de-bruijn-graphs-hamiltonian-paths-and-eulerian-paths.html">Part 1 of this post</a>:</p>
<ul>
<li>An Eulerian path is a path that visits every edge of a given graph exactly once.</li>
<li>An Eulerian cycle is an Eulerian path that begins and ends at the ''same vertex''.</li>
</ul>
<p>According to Steven Skienna's <u>Algorithm Design Handbook</u>,
there are two conditions that must be met for an Eulerian path or
cycle to exist. These conditions are different for undirected graphs
versus directed graphs.</p>
<p><strong>Undirected graphs:</strong> </p>
<ul>
<li>
<p>An undirected graph contains an Euler <strong>cycle</strong> iff
  (1) it is connected, and (2) each vertex is of even
  degree.</p>
</li>
<li>
<p>An undirected graph contains an Euler <strong>path</strong> iff (1)
  it is connected, and all but two vertices are of even
  degree. These two vertices will be the start and end
  vertices for the Eulerian path.</p>
</li>
</ul>
<p><strong>Directed graphs:</strong></p>
<ul>
<li>
<p>A directed graph contains an Euler cycle iff (1) it is
  strongly-connected, and (2) each vertex has the same
  in-degree as out-degree</p>
</li>
<li>
<p>A directed graph contains an Euler path iff (1) it is
  connected, and (2) all vertices except two (x,y) have
  the same in-degree as out-degree, and (x,y) are
  vertices with in-degree one less than and one more
  than out-degree</p>
</li>
</ul>
<h1>Algorithm Building Blocks</h1>
<p>Algorithm to find Eulerian paths/cycles consists of several steps using several algorithms.</p>
<p>Undirected graphs are the simpler case; directed graphs are more complicated.</p>
<h2>Depth First Search</h2>
<p>To perform a DFS on a directed graph, implement two functions:</p>
<ol>
<li>
<p>Write a DFS function that takes a graph as an input argument and that visits each node of the
   graph in a depth-first search.</p>
</li>
<li>
<p>Write a visitation function that takes a node as an input argument and that performs some action
   on the node. This visitation function is called by the DFS function on each node that it visits.</p>
</li>
</ol>
<h2>Kosaraju's Algorithm: Connected Components</h2>
<p>On an undirected graph, can use Fleury's Algorithm
to follow edges (classify edges as bridge or non-bridge,
then leave bridges for last).</p>
<p>On a directed graph, we have twice the amount of work:
we are not just checking that all vertices are reachable
from a given vertex, we are also checking that all vertices
can also reach that vertex.</p>
<h1>To Be Continued...</h1>
<p>In the next part of this post, we will start with the slightly simpler case of finding an Euler cycle
(which has no start or end vertices). Then we will show how finding the Euler path is actually a special
case of finding the Euler cycle.</p>
<p>First, we will use <a href="https://charlesreid1.com/wiki/Graphs/Euler_Circuit#Directed_Graphs:_Hierholzer.27s_Algorithm">Hierholzer's
Algorithm</a>
to find Euler cycles (this is the simpler case). Order does not matter because it is a cycle;
Hierholzer's algorithm is used to find the Euler cycle.</p>
<p>Next, we will modify the above algorithm to find Euler paths. This requires keeping track of
the start and end candidate nodes. We verify only one each; we complete the cycle by adding an edge.
Once we find the cycle, we remove the edge. Finally, we rearrange the cycle to have the correct
start and end nodes.</p>
<p>Stay tuned for Part 3...</p>
<script type="text/javascript">if (!document.getElementById('mathjaxscript_pelican_#%@#$@#')) {
    var align = "center",
        indent = "0em",
        linebreak = "false";

    if (false) {
        align = (screen.width < 768) ? "left" : align;
        indent = (screen.width < 768) ? "0em" : indent;
        linebreak = (screen.width < 768) ? 'true' : linebreak;
    }

    var mathjaxscript = document.createElement('script');
    mathjaxscript.id = 'mathjaxscript_pelican_#%@#$@#';
    mathjaxscript.type = 'text/javascript';
    mathjaxscript.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.3/latest.js?config=TeX-AMS-MML_HTMLorMML';
    mathjaxscript[(window.opera ? "innerHTML" : "text")] =
        "MathJax.Hub.Config({" +
        "    config: ['MMLorHTML.js']," +
        "    TeX: { extensions: ['AMSmath.js','AMSsymbols.js','noErrors.js','noUndefined.js'], equationNumbers: { autoNumber: 'AMS' } }," +
        "    jax: ['input/TeX','input/MathML','output/HTML-CSS']," +
        "    extensions: ['tex2jax.js','mml2jax.js','MathMenu.js','MathZoom.js']," +
        "    displayAlign: '"+ align +"'," +
        "    displayIndent: '"+ indent +"'," +
        "    showMathMenu: true," +
        "    messageStyle: 'normal'," +
        "    tex2jax: { " +
        "        inlineMath: [ ['\\\\(','\\\\)'] ], " +
        "        displayMath: [ ['$$','$$'] ]," +
        "        processEscapes: true," +
        "        preview: 'TeX'," +
        "    }, " +
        "    'HTML-CSS': { " +
        "        styles: { '.MathJax_Display, .MathJax .mo, .MathJax .mi, .MathJax .mn': {color: 'inherit ! important'} }," +
        "        linebreaks: { automatic: "+ linebreak +", width: '90% container' }," +
        "    }, " +
        "}); " +
        "if ('default' !== 'default') {" +
            "MathJax.Hub.Register.StartupHook('HTML-CSS Jax Ready',function () {" +
                "var VARIANT = MathJax.OutputJax['HTML-CSS'].FONTDATA.VARIANT;" +
                "VARIANT['normal'].fonts.unshift('MathJax_default');" +
                "VARIANT['bold'].fonts.unshift('MathJax_default-bold');" +
                "VARIANT['italic'].fonts.unshift('MathJax_default-italic');" +
                "VARIANT['-tex-mathit'].fonts.unshift('MathJax_default-italic');" +
            "});" +
            "MathJax.Hub.Register.StartupHook('SVG Jax Ready',function () {" +
                "var VARIANT = MathJax.OutputJax.SVG.FONTDATA.VARIANT;" +
                "VARIANT['normal'].fonts.unshift('MathJax_default');" +
                "VARIANT['bold'].fonts.unshift('MathJax_default-bold');" +
                "VARIANT['italic'].fonts.unshift('MathJax_default-italic');" +
                "VARIANT['-tex-mathit'].fonts.unshift('MathJax_default-italic');" +
            "});" +
        "}";
    (document.body || document.getElementsByTagName('head')[0]).appendChild(mathjaxscript);
}
</script></div>

    </div>
</div>

<div class="v20"></div>

<div class="row">
    <div class="col-md-12" style="text-align: left;">

        <div class="tags">
            <p>Tags:&nbsp;&nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/go.html">go</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/golang.html">golang</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/rosalind.html">rosalind</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/computational-biology.html">computational biology</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/bioinformatics.html">bioinformatics</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/euler.html">euler</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/recursion.html">recursion</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/backtracking.html">backtracking</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/graphs.html">graphs</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/algorithms.html">algorithms</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/hamiltonian.html">hamiltonian</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/eulerian.html">eulerian</a>
                    &nbsp;&nbsp;
        </div>

    </div>
</div>


<div class="v200"></div>
  	            	</article>
  	            	<article>
<header>
<div class="row">


    <div class="col-md-12">
        <div class="headtext" style="text-align: center;">
            <h1><b>
                Graphs for Bioinformatics, Part 1: de Bruijn Graphs, Hamiltonian Paths, and Eulerian Paths
            </b></h1>
        
            <p class="lead">
                Posted 
                <time datetime="2019-05-02T19:00:00-07:00" pubdate>Thursday 05/02/2019</time>
                in 
                <a href="https://charlesreid1.github.io/category/computational-biology.html">Computational Biology</a>
                </p>
            <p>
            <span style="font-size: 14px">
                <a href="https://charlesreid1.github.io/graphs-for-bioinformatics-part-1-de-bruijn-graphs-hamiltonian-paths-and-eulerian-paths.html">permalink</a>
            </span>
            </p>
        </div>

    </div>

</div>
</header>


<div class="v50"></div>


<div class="row">
    <div class="col-md-12" style="text-align: left;">

        <div class="entry-content"><h1>Table of Contents</h1>
<ul>
<li><a href="#the-context-rosalindinfo">The Context: Rosalind.info</a></li>
<li><a href="#graphs-for-bioinformatics">Graphs for Bioinformatics</a><ul>
<li><a href="#building-a-k-mer-graph-the-wrong-graph">Building a K-mer Graph (The Wrong Graph)</a></li>
<li><a href="#building-a-de-bruijn-graph-the-right-graph">Building a De Bruijn Graph (The Right Graph)</a></li>
<li><a href="#transform-the-problem-hamiltonian-paths-to-eulerian-paths">Transform the Problem: Hamiltonian Paths to Eulerian Paths</a></li>
</ul>
</li>
<li><a href="#an-example">An Example</a></li>
<li><a href="#back-to-dna">Back to DNA</a></li>
</ul>
<p><br />
<br /></p>
<h1>The Context: Rosalind.info</h1>
<p>To provide a bit of context for a discussion of Euler paths and Euler cycles:
starting around December, a group of us in the <a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology (DIB Lab)</a>
started working through the textbook <a href="http://bioinformaticsalgorithms.com/">Bioinformatics Algorithms: An Active Learning Approach</a>
and the associated website, <a href="https://rosalind.info">Rosalind.info</a>.</p>
<p>Rosalind.info is a site that is similar in style to <a href="https://projecteuler.net/">Project Euler</a>,
a <a href="https://charlesreid1.github.io/tag/project-euler.html">familiar topic on this blog</a>.
Project Euler poses computationally challenging problems in the domain of mathematics.</p>
<p>Like Project Euler, the visitor is given one small example input and the corresponding 
correct output, and one large example input and corresponding output. Also like Project
Euler, the problems vary in how much computer science versus domain expertise is needed,
but they are largely focused on writing algorithms rather than on the science behind the
computations.</p>
<p>Unlike Project Euler, however, Rosalind.info does give plenty of hints (via the textbook,
if you have a copy), and sometimes even gives pseudocode for the algorithm. <strong>The book is
required to get enough context to answer some of the Rosalind.info problems.</strong></p>
<h1>Graphs for Bioinformatics</h1>
<p>The textbook focuses on different problems in each chapter. For example, Chapter 1 uses
the example of a string of DNA that marks where replication begins to introduce some
basic bioinformatics concepts and algorithms. Chapter 2 uses the concept of molecular
clocks to introduce motifs and motif-finding, the focus of most of the problems in
Chapter 2.</p>
<p>Chapter 3 focuses on the problem of genome assembly - how we assemble an entire genome
from short segments alone. In particular, the chapter focuses on de Bruijn graphs, 
which are graphs that, given a sequence of symbols drawn from an alphabet, are composed 
of edges (one for each <strong>k-mer</strong>, that is, a chunk of the sequence of length k),
and vertices (one for each k-mer prefix and k-mer suffix, connected by a directed edge
of the k-mer). We will cover more of the details of these graphs shortly.</p>
<h2>Building a K-mer Graph (The Wrong Graph)</h2>
<p>The Bioinformatics Algorithm book starts with a general discussion of how to 
represent a sequence of DNA nucleotides using a graph. The idea they discuss 
initially (which is an obvious, but not necessarily good, one) is splitting
the sequence into k-mer chunks, like so:</p>
<div class="highlight"><pre><span></span>      Sequence:   TAATGCCATGGGATGTT
      Pieces:     TAA 
                   AAT
                    ATG 
                     TGC
                      GCC 
                       CCA
                        CAT 
                         ATG
                          TGG 
                           GGG
                            GGA 
                             GAT
                              ATG 
                               TGT
                                GTT
</pre></div>


<p>and letting one k-mer be represented by one vertex. Then the sequence
above could be turned into the graph:</p>
<div class="highlight"><pre><span></span>TAA -&gt; AAT -&gt; ATG -&gt; TGC -&gt; GCC -&gt; CCA -&gt; CAT -&gt; ATG -&gt; TGG -&gt; GGG -&gt; GGA -&gt; GAT -&gt; ATG -&gt; TGT -&gt; GTT
</pre></div>


<p>On this graph, every edge has the property that the first (k-1)
nucleotides of the destination match the last (k-1) nucleotides
of the source. </p>
<p>If we did not know this sequence in advance, we could draw <em>every</em>
edge with that property - every time the last (k-1) characters of
a k-mer match the first (k-1) characters of another k-mer, an edge
is drawn between those two vertices. </p>
<p>That graph would result in <em>many</em> more edges than the graph shown above.</p>
<p>Furthermore, in theory, if each read sequence came from a single genome
and we had the entire genome covered by read sequences, a path through
the graph that visits every vertex (every k-mer) would yield the full
genome.</p>
<p>A path through a graph that visits every vertex once is called a 
<strong>Hamiltonian path</strong>.</p>
<p>Why is this hard? Because the problem of proving a Hamiltonian
path exists, let alone finding it, becomes very difficult for
large graphs.</p>
<h2>Building a De Bruijn Graph (The Right Graph)</h2>
<p>Nicolaas de Bruijn introduced (in 1946, in a paper entitled simply
<a href="https://pure.tue.nl/ws/files/4442708/597473.pdf">"A combinatorial problem"</a>)
a new way of representing a sequence with a graph. He split a given
sequence into k-mers, as before, but instead of representing
each k-mer as a <em>vertex</em> on the graph, he represented each 
k-mer as an <em>edge</em> on the graph.</p>
<p>This type of graph is called a <strong>de Bruijn graph</strong>.</p>
<p>Specifically, for a DNA sequence, each k-mer from the sequence is
represented by an edge, where the source vertex
is that k-mer's (k-1)-nucleotide suffix and the destination vertex
is that k-mer's (k-1)-nucleotide prefix.</p>
<div class="highlight"><pre><span></span>      Sequence:   TAATGCCATGGGATGTT
      Pieces:     TA  
                   AA
                    AT 
                     TG
                      GC 
                       CC
                        CA 
                         AT
                          TG 
                           GG
                            GG 
                             GA 
                              AT  
                               TG 
                                GT 
                                 TT
</pre></div>


<p>Now this sequence is written as the graph:</p>
<div class="highlight"><pre><span></span>TA -&gt; AA -&gt; AT -&gt; TG -&gt; GC -&gt; CC -&gt; CA -&gt; AT -&gt; TG -&gt; GG -&gt; GG -&gt; GA -&gt; AT -&gt; TG -&gt; GT -&gt; TT
</pre></div>


<p>so that the original breakup of the sequence into k-mers is still
represented, but now as edges rather than as vertices. That is, the 
k-mer <code>TAA</code> is represented by the edge <code>TA -&gt; AA</code>.</p>
<h2>Transform the Problem: Hamiltonian Paths to Eulerian Paths</h2>
<p>The change in the problem representation (k-mers as vertices to k-mers
as edges) changes the problem of finding the <strong>Hamiltonian path</strong> 
(a path through the graph that visits every <em>vertex</em> exactly once) 
into the problem of finding the <strong>Eulerian path</strong>
(a path through the graph that visits every <em>edge</em> exactly once).</p>
<h1>An Example</h1>
<p>Let's look at a slightly simpler example - the one de Bruijn was
originally considering - so we can see de Bruijn graphs in action
in a slightly simpler case.</p>
<p>In his 1946 paper <a href="https://pure.tue.nl/ws/files/4442708/597473.pdf">"A combinatorial problem"</a>,
de Bruijn describes the problem thus:</p>
<blockquote>
<p>Some years ago Ir. K. Posthumus stated an interesting conjecture
concerning certain cycles of digits 0 or 1, which we shall call
<span class="math">\(P_n\)</span> cycles. For <span class="math">\(n = 1, 2, 3, \dots\)</span>, a <span class="math">\(P_n\)</span> cycle be an ordered
cycle of <span class="math">\(2^n\)</span> digits 0 or 1, such that the <span class="math">\(2^n\)</span> possible ordered
sets of <span class="math">\(n\)</span> consecutive digits of that cycle are all different.
As a consequence, any ordered set of <span class="math">\(n\)</span> digits 0 or 1 occurs exactly
once in that cycle.</p>
<p>For example, a <span class="math">\(P_3\)</span> cycle is <span class="math">\(00010111\)</span>, respectively showing the
triples 000, 001, 010, 011, 111, 100, 100, which are all the possible
triples indeed.</p>
</blockquote>
<p>In this case, de Bruijn is discussing <em>complete</em> de Bruijn graphs - he
constructs a de Bruijn graph of all possible 3-mers (our k-mers, <span class="math">\(k = 3\)</span>),
and constructs a path through the graph that visits every edge of the
graph. Here is the sequence broken down as above:</p>
<div class="highlight"><pre><span></span>      Sequence:   00010111
      Pieces:     00  
                   00
                    01
                     10
                      01
                       11
                        11
</pre></div>


<p>The alphabet here is binary: 0 and 1. </p>
<p>This (seemingly simple) example is a bit confusing, but here's
what's going on: we have four vertices on the de Bruijn graph,
consisting of the 2-mers:</p>
<div class="highlight"><pre><span></span>00
01
10
11
</pre></div>


<p>Now, if we draw an edge for every possible 3-mer, we would 
start with the 3-mer <code>000</code>, which is actually represented
by a self-edge from vertex <code>00</code> to vertex <code>00</code>, because the
prefix matches the suffix.</p>
<p>Similarly, the 3-mer <code>111</code> is represented by a self-edge
from vertex <code>11</code> to vertex <code>11</code>.</p>
<p>The other 3-mers are represented by their corresponding edges:
<code>001</code> is represented by the edge <code>00 -&gt; 01</code>, <code>010</code> by the edge
<code>01 -&gt; 10</code>, etc.</p>
<p>By drawing <strong>every possible edge</strong> (to represent every possible
3-mer), we assemble the <strong>complete de Bruijn graph</strong> (that is,
the de Bruijn graph containing vertices for all possible
2-mers connected by edges representing every possible 3-mer
in the given alphabet).</p>
<p>The sequence de Bruijn gives in his paper is an Euler path
through the <em>complete</em> (de Bruijn) graph (that is, a path 
through the de Bruijn graph that visits every edge exactly 
once):</p>
<div class="highlight"><pre><span></span><span class="n">Sequence</span><span class="o">:</span> <span class="mi">00010111</span>

<span class="mi">00</span> <span class="o">-&gt;</span> <span class="mi">00</span> <span class="o">-&gt;</span> <span class="mi">01</span> <span class="o">-&gt;</span> <span class="mi">10</span> <span class="o">-&gt;</span> <span class="mi">01</span> <span class="o">-&gt;</span> <span class="mi">11</span> <span class="o">-&gt;</span> <span class="mi">11</span>
</pre></div>


<h1>Back to DNA</h1>
<p>Now the utility of the de Bruijn methodology is more clear:
if we can come up with fast, efficient algorithms to find
Euler paths on large graphs, we can transform the assembly 
problem (given fragments of a long sequence, reconstruct
the sequence) into the problem of finding an Eulerian path,
which is tractable even for large graphs.</p>
<p>Compare this with string matching algorithms utilizing
dynamic programming, which can cost <span class="math">\(O(N^2)\)</span> and make
genome assembly computationally infeasible.</p>
<p>Part 1 (this post) has covered the basic idea behind 
assembling DNA sequences using de Bruijn graphs. In
Part 2 of this post, we will move on to a discussion
of the "real world" problem, warts and all, and how 
we can relax some of the strict assumptions (like assuming
perfect coverage of the original sequence and assuming
that all reads are perfect).</p>
<script type="text/javascript">if (!document.getElementById('mathjaxscript_pelican_#%@#$@#')) {
    var align = "center",
        indent = "0em",
        linebreak = "false";

    if (false) {
        align = (screen.width < 768) ? "left" : align;
        indent = (screen.width < 768) ? "0em" : indent;
        linebreak = (screen.width < 768) ? 'true' : linebreak;
    }

    var mathjaxscript = document.createElement('script');
    mathjaxscript.id = 'mathjaxscript_pelican_#%@#$@#';
    mathjaxscript.type = 'text/javascript';
    mathjaxscript.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.3/latest.js?config=TeX-AMS-MML_HTMLorMML';
    mathjaxscript[(window.opera ? "innerHTML" : "text")] =
        "MathJax.Hub.Config({" +
        "    config: ['MMLorHTML.js']," +
        "    TeX: { extensions: ['AMSmath.js','AMSsymbols.js','noErrors.js','noUndefined.js'], equationNumbers: { autoNumber: 'AMS' } }," +
        "    jax: ['input/TeX','input/MathML','output/HTML-CSS']," +
        "    extensions: ['tex2jax.js','mml2jax.js','MathMenu.js','MathZoom.js']," +
        "    displayAlign: '"+ align +"'," +
        "    displayIndent: '"+ indent +"'," +
        "    showMathMenu: true," +
        "    messageStyle: 'normal'," +
        "    tex2jax: { " +
        "        inlineMath: [ ['\\\\(','\\\\)'] ], " +
        "        displayMath: [ ['$$','$$'] ]," +
        "        processEscapes: true," +
        "        preview: 'TeX'," +
        "    }, " +
        "    'HTML-CSS': { " +
        "        styles: { '.MathJax_Display, .MathJax .mo, .MathJax .mi, .MathJax .mn': {color: 'inherit ! important'} }," +
        "        linebreaks: { automatic: "+ linebreak +", width: '90% container' }," +
        "    }, " +
        "}); " +
        "if ('default' !== 'default') {" +
            "MathJax.Hub.Register.StartupHook('HTML-CSS Jax Ready',function () {" +
                "var VARIANT = MathJax.OutputJax['HTML-CSS'].FONTDATA.VARIANT;" +
                "VARIANT['normal'].fonts.unshift('MathJax_default');" +
                "VARIANT['bold'].fonts.unshift('MathJax_default-bold');" +
                "VARIANT['italic'].fonts.unshift('MathJax_default-italic');" +
                "VARIANT['-tex-mathit'].fonts.unshift('MathJax_default-italic');" +
            "});" +
            "MathJax.Hub.Register.StartupHook('SVG Jax Ready',function () {" +
                "var VARIANT = MathJax.OutputJax.SVG.FONTDATA.VARIANT;" +
                "VARIANT['normal'].fonts.unshift('MathJax_default');" +
                "VARIANT['bold'].fonts.unshift('MathJax_default-bold');" +
                "VARIANT['italic'].fonts.unshift('MathJax_default-italic');" +
                "VARIANT['-tex-mathit'].fonts.unshift('MathJax_default-italic');" +
            "});" +
        "}";
    (document.body || document.getElementsByTagName('head')[0]).appendChild(mathjaxscript);
}
</script></div>

    </div>
</div>

<div class="v20"></div>

<div class="row">
    <div class="col-md-12" style="text-align: left;">

        <div class="tags">
            <p>Tags:&nbsp;&nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/go.html">go</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/golang.html">golang</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/rosalind.html">rosalind</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/computational-biology.html">computational biology</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/bioinformatics.html">bioinformatics</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/euler.html">euler</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/recursion.html">recursion</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/backtracking.html">backtracking</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/graphs.html">graphs</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/algorithms.html">algorithms</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/hamiltonian.html">hamiltonian</a>
                    &nbsp;&nbsp;
                    <a href="https://charlesreid1.github.io/tag/eulerian.html">eulerian</a>
                    &nbsp;&nbsp;
        </div>

    </div>
</div>


<div class="v200"></div>
  	            	</article>

<div class="row">
    <div class="col-md-4 col-centered" style="text-align: center;">

        <div class="pagination">

            <a class="next" href="https://charlesreid1.github.io/index2.html">&larr; Newer</a>

            &nbsp;
            &nbsp;
            &nbsp;
            &nbsp;
            &nbsp;

            <a class="prev" href="https://charlesreid1.github.io/index4.html">Older &rarr;</a>
        
          <br />
        </div>

    </div>
</div>

<div class="v50"></div>


        </div>

        <div class="col-md-4" style="text-align: center;">
<a name="a_archive"></a>


        <div class="well">

            <h2>December 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/creating-mock-api-servers.html">Creating Mock API Servers</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/automatically-generating-up-to-date-requirementstxt-for-python-projects.html">Automatically Generating Up-To-Date requirements.txt for Python Projects</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/git-workflows-part-3-refactoring-large-branches-and-pull-requests.html">Git Workflows, Part 3: Refactoring Large Branches and Pull Requests</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>November 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/git-workflows-part-2-crafting-commits.html">Git Workflows, Part 2: Crafting Commits</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>October 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/git-workflows-part-1-supercharging-your-git-config.html">Git Workflows, Part 1: Supercharging your Git Config</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>September 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/mocking-aws-in-unit-tests.html">Mocking AWS in Unit Tests</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>May 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/deconvoluting-convolutional-neural-networks.html">Deconvoluting Convolutional Neural Networks</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/graphs-for-bioinformatics-part-2-finding-eulerian-paths.html">Graphs for Bioinformatics, Part 2: Finding Eulerian Paths</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/graphs-for-bioinformatics-part-1-de-bruijn-graphs-hamiltonian-paths-and-eulerian-paths.html">Graphs for Bioinformatics, Part 1: de Bruijn Graphs, Hamiltonian Paths, and Eulerian Paths</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>April 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/the-git-commit-ectomy.html">The Git-Commit-Ectomy</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>March 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/the-josephus-problem-part-3-solving-the-double-step-case.html">The Josephus Problem: Part 3: Solving the Double Step Case</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/the-josephus-problem-part-2-two-examples.html">The Josephus Problem: Part 2: Two Examples</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/the-josephus-problem-part-1-the-problem.html">The Josephus Problem: Part 1: The Problem</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/approximating-pi-happy-pi-day.html">Approximating Pi (Happy Pi Day)</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/five-letter-words-part-5-the-try-trie-tree.html">Five Letter Words: Part 5: The Try Trie Tree</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/five-letter-words-part-4-revisiting-diff-by-one.html">Five Letter Words: Part 4: Revisiting Diff by One</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/centillion-a-document-search-engine.html">centillion: a document search engine</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>February 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/any-color-you-like-as-long-as-its-00add8.html">Any Color You Like, As Long As It's 00ADD8</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/a-few-of-my-favorite-peps.html">A Few of My Favorite PEPs</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/context-managers-in-python.html">Context Managers in Python</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>January 2019</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/building-snakemake-command-line-wrappers-for-kubernetes-workflows.html">Building Snakemake Command Line Wrappers for Kubernetes Workflows</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/building-snakemake-command-line-wrappers-for-workflows.html">Building Snakemake Command Line Wrappers for Workflows</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/recursive-backtracking-in-go-for-bioinformatics-applications-3-go-implementation-of-backtracking.html">Recursive Backtracking in Go for Bioinformatics Applications: 3. Go Implementation of Backtracking</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>December 2018</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/recursive-backtracking-in-go-for-bioinformatics-applications-2-generating-variations.html">Recursive Backtracking in Go for Bioinformatics Applications: 2. Generating Variations</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/recursive-backtracking-in-go-for-bioinformatics-applications-1-counting-variations.html">Recursive Backtracking in Go for Bioinformatics Applications: 1. Counting Variations</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/basic-data-structures-in-go-maps.html">Basic Data Structures in Go: Maps</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/learning-bioinformatics-with-go-and-rosalind.html">Learning Bioinformatics with Go and Rosalind</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>October 2018</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/first-post-of-the-fall-part-2-flaskadillo.html">First Post of the Fall, Part 2: Flaskadillo</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/first-post-of-the-fall-part-1-data-commons.html">First Post of the Fall, Part 1: Data Commons</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>May 2018</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/current-projects.html">Current Projects</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>March 2018</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/charlesreid1com-stack.html">Charlesreid1.com Stack</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>February 2018</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/d3-calendar-visualizations.html">D3 Calendar Visualizations</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/project-euler-problem-172.html">Project Euler Problem 172</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>January 2018</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/4x4-rubiks-cube-part-4-sequence-order.html">4x4 Rubik's Cube: Part 4: Sequence Order</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/4x4-rubiks-cube-part-3-factoring-permutations.html">4x4 Rubik's Cube: Part 3: Factoring Permutations</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/4x4-rubiks-cube-part-2-permutations.html">4x4 Rubik's Cube: Part 2: Permutations</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/4x4-rubiks-cube-part-1-representations.html">4x4 Rubik's Cube: Part 1: Representations</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/lets-generate-permutations.html">Let's Generate Permutations!</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>September 2017</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/five-letter-words-part-3-letter-coverage-and-dynamic-programming.html">Five Letter Words: Part 3: Letter Coverage and Dynamic Programming</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/five-letter-words-part-2-more-five-word-algorithms.html">Five Letter Words: Part 2: More Five-Word Algorithms</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/five-letter-words-part-1-getting-familiar-with-the-list.html">Five Letter Words: Part 1: Getting Familiar With The List</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>August 2017</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/eulers-theorem-the-totient-function-and-calculating-totients-by-hand.html">Euler's Theorem, the Totient Function, and Calculating Totients By Hand</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>July 2017</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/mad-combinatoric-castles.html">Mad Combinatoric Castles</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/project-euler-problem-1.html">Project Euler Problem 1</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/shortest-lattice-paths-and-multiset-permutations.html">Shortest Lattice Paths and Multiset Permutations</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/computing-square-roots-part-2-using-continued-fractions.html">Computing Square Roots: Part 2: Using Continued Fractions</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/computing-square-roots-part-1-using-newtons-method.html">Computing Square Roots: Part 1: Using Newton's Method</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>June 2017</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/cse-143-final-project-hilbert-sort-3-the-code.html">CSE 143 Final Project: Hilbert Sort: 3. The Code</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/cse-143-final-project-hilbert-sort-2-the-solution-algorithm.html">CSE 143 Final Project: Hilbert Sort: 2. The Solution Algorithm</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/cse-143-final-project-hilbert-sort-1-the-problem.html">CSE 143 Final Project: Hilbert Sort: 1. The Problem</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/cse-143-final-project-classy.html">CSE 143 Final Project: Classy</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/cse-143-final-project-checkers.html">CSE 143 Final Project: Checkers</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>May 2017</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/teaching-recursion-with-the-n-queens-problem.html">Teaching Recursion with the N Queens Problem</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/undergraduate-research-project-wireless-sensor-networks-for-internet-of-things-applications-part-2-the-technologies.html">Undergraduate Research Project: Wireless Sensor Networks for Internet of Things Applications (Part 2: The Technologies)</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/undergraduate-research-project-wireless-sensor-networks-for-internet-of-things-applications-part-1-the-project.html">Undergraduate Research Project: Wireless Sensor Networks for Internet of Things Applications (Part 1: The Project)</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>April 2017</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/stunnel.html">Stunnel</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/traveling-schoolteacher-problem.html">Traveling Schoolteacher Problem</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/the-z-machine-a-simple-turing-machine.html">The Z-Machine: A Simple Turing Machine</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/awsome-day-seattle-notes-part-2-networking-security-and-miscellany.html">AWSome Day Seattle Notes: Part 2: Networking, Security, and Miscellany</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/awsome-day-seattle-notes-part-1-the-basics.html">AWSome Day Seattle Notes: Part 1: The Basics</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/setting-up-a-self-hosted-github-clone-with-gitea.html">Setting Up a Self-Hosted Github Clone with Gitea</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/better-timing-of-guava-traveling-salesperson-problem-code-timing-scripts.html">Better Timing of Guava Traveling Salesperson Problem Code: Timing Scripts</a>
            </p>

        </div>

        <div class="v20"></div>


        <div class="well">

            <h2>March 2017</h2>

            <p class="archive">
            <a href="https://charlesreid1.github.io/fixing-bottlenecks-in-the-guava-traveling-salesperson-problem-code.html">Fixing Bottlenecks in the Guava Traveling Salesperson Problem Code</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/python-vs-perl-n-queens-problem.html">Python vs. Perl: N Queens Problem</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/solving-the-traveling-salesperson-problem-with-java-and-guava.html">Solving the Traveling Salesperson Problem with Java and Guava</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/perl-vs-java-n-queens-problem.html">Perl vs. Java: N Queens Problem</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/enigma-cipher-implementation-part-4-combinatorics.html">Enigma Cipher Implementation: Part 4: Combinatorics</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/enigma-cipher-implementation-part-3-enigma-in-java-without-objects.html">Enigma Cipher Implementation: Part 3: Enigma in Java Without Objects</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/enigma-cipher-implementation-part-2-pseudocode.html">Enigma Cipher Implementation: Part 2: Pseudocode</a>
            </p>
            <p class="archive">
            <a href="https://charlesreid1.github.io/enigma-cipher-implementation-part-1-how-it-works.html">Enigma Cipher Implementation: Part 1: How It Works</a>
            </p>

        </div>

        <div class="v20"></div>



<div class="v100"></div>
        </div>

    </div>



</div>



<script type="text/javascript" src="https://charlesreid1.github.io/theme/js/jquery-1.11.2.js"></script>
<script type="text/javascript" src="https://charlesreid1.github.io/theme/js/bootstrap-3.3.4.js"></script>
<script type="text/javascript" src="https://charlesreid1.github.io/theme/js/d3-3.5.5.js"></script><script type="text/javascript" src="https://charlesreid1.github.io/theme/js/jquery-1.11.2.js"></script>
<script type="text/javascript" src="https://charlesreid1.github.io/theme/js/bootstrap-3.3.4.js"></script>

<footer id="contentinfo" class="body">
<p>&nbsp;</p>
<p>&nbsp;</p>
<center>
	<hr />
</center>
<p>&nbsp;</p>
<p style="text-align: center">
    <span class="fa-stack fa-lg">
        <i class="fa fa-square fa-stack-2x" style="color:#000;"></i>
        <i class="fa fa-terminal fa-stack-1x fa-inverse"></i>
    </span>
    Made from the command line with vim by 
    <a href="http://charlesreid1.com">charlesreid1</a><br />
    with help from <a href="https://getbootstrap.com/">Bootstrap</a> and <a href="http://getpelican.com">Pelican</a>.
</p>

<br />

<p style="text-align: center">
    <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">
    <span class="fa-stack fa-lg">
        <i class="fa fa-square fa-stack-2x" style="color:#000;"></i>
        <i class="fa fa-creative-commons fa-stack-1x fa-inverse"></i>
    </span>
    </a>
    <br />
    Licensed under the <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 License</a>.
</p>

<!-- Start of StatCounter Code for Default Guide -->
<script type="text/javascript">
var sc_project=11611748;
var sc_invisible=1;
var sc_security="9fcba820";
</script>
<script type="text/javascript"
src="https://www.statcounter.com/counter/counter.js"
async></script>

<noscript><div class="statcounter"><a title="Web Analytics"
href="http://statcounter.com/" target="_blank"><img
class="statcounter"
src="//c.statcounter.com/11611748/0/9fcba820/1/" alt="Web
Analytics"></a></div></noscript>
<!-- End of StatCounter Code for Default Guide -->
</footer><!-- /#contentinfo -->
</body>
</html>