<!DOCTYPE html>
<html lang="en">
<head>
        <title>charlesreid1</title>
        <meta charset="utf-8" />

        <!--
        CSS styles
        -->
        <link href="/theme/css/bootstrap.css"       rel="stylesheet" type="text/css">
        <link href="/theme/css/slate.css"              rel="stylesheet" type="text/css">
        <link href="/theme/css/font-awesome.css"       rel="stylesheet" type="text/css"/>
        <link href="/theme/css/pygment-solarized.css"  rel="stylesheet" type="text/css"/>

        <!--
        my CSS styles
        -->
        <link href="/theme/css/main.css"            rel="stylesheet" type="text/css">
        <link href="/theme/css/dox.css"             rel="stylesheet" type="text/css">


        <!--
        include Angular first
        -->
        <script type="text/javascript" src="/theme/js/angular-1.3.15.js"></script>
        <script type="text/javascript" src="/theme/js/skrollr-0.6.29.js"></script>
        <script type="text/javascript" src="/theme/js/d3-3.5.5.js"></script>

        <!--
        this seems like a bad idea.
        but if I don't put jquery first, I get errors about unrecognized $s.
        -->
        <script type="text/javascript" src="/theme/js/jquery-1.11.2.js"></script>

</head>
<body id="index">
<nav class="navbar navbar-default">
    <div class="container-fluid">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#myNavbar">
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span> 
            </button>
            <a href="https://charlesreid1.github.io" class="navbar-brand">charlesreid1.github.io</a>
        </div>
        <div>
            <div class="collapse navbar-collapse" id="myNavbar">
                <ul class="nav navbar-nav">

                    <li>
                        <a href="https://charlesreid1.com/wiki">Wiki</a>
                    </li>
                    <li>
                        <a href="https://git.charlesreid1.com/explore">Git</a>
                    </li>
                    <li>
                        <a href="http://spotify.charlesreid1.com">Music</a>
                    </li>
                    <li>
                        <a href="https://charlesreid1.com/life">Life</a>
                    </li>
                    <li>
                        <a href="https://charlesreid1.github.io">Blog</a>
                    </li>
                    <li>
                        <a href="https://charlesreid1.com/about">About Me</a>
                    </li>

                </ul>
            </div>
        </div>
    </div>
</nav>



<div class="container">

<header>
<div class="row">


    <div class="col-md-12">
        <div class="headtext" style="text-align: center;">
            <h1><b>
                Recursive Backtracking in Go for Bioinformatics Applications: 2. Generating Variations
            </b></h1>
        
            <p class="lead">
                Posted 
                <time datetime="2018-12-27T18:00:00-08:00" pubdate>Thursday 12/27/2018</time>
                in 
                <a href="/category/rosalind.html">Rosalind</a>
                </p>
            <p>
            <span style="font-size: 14px">
                <a href="/drafts/recursive-backtracking-in-go-for-bioinformatics-applications-2-generating-variations.html">permalink</a>
            </span>
            </p>
        </div>

    </div>

</div>
</header>


<div class="v50"></div>


<div class="row">
    <div class="col-md-12" style="text-align: left;">

        <div class="entry-content"><p><em>This is the second in a series of three blog posts describing our
solution to a bioinformatics problem from Rosalind.info,
<a href="http://rosalind.info/problems/ba1i/">Problem BA1(i) (Find most frequent words with mismatches
in a string)</a>.
To solve this problem and generate variations of a DNA string
as required, we implemented a recursive backtracking method
in the Go programming language.</em></p>
<ul>
<li><em><a href="#">Part 1: Counting Variations</a></em></li>
<li><em>Part 2: Generating Variations (you are here)</em></li>
<li><em><a href="#">Part 3: Go Implementation of Recursive Backtracking</a></em></li>
</ul>
<p><br />
<br /></p>
<p><a name="toc"></a></p>
<h2>Table of Contents</h2>
<ul>
<li><a href="#problem-descr">Problem Description</a></li>
<li><a href="#perms-combs-vars">Permutations vs Combinations vs Variations</a></li>
<li><a href="#transform">Transforming the Problem Space</a><ul>
<li><a href="#binary">Generating Binary Numbers With Constraints</a></li>
<li><a href="#bitvectors">Bitvectors</a></li>
<li><a href="#recursion">Recursion</a></li>
</ul>
</li>
<li><a href="#backtracking">Recursive Backtracking Pseudocode</a></li>
<li><a href="#dna">Appying to DNA Variations</a></li>
</ul>
<p><br />
<br /></p>
<p><a name="problem-descr"></a></p>
<h2>Problem Description</h2>
<p>The task at hand is to take a given input strand of DNA,
and generate variations from it that have up to <span class="math">\(d\)</span> differences
(a Hamming distance of <span class="math">\(d\)</span>) in the codons (base pairs).</p>
<p>In <a href="#">part 1 of this series</a>, we walk through the
construction of an analytical formula to count the number
of variations of a given DNA string that can be generated,
given the constraints of the problem.</p>
<p>In <a href="#">part 2 of this series</a>, we cover several techniques to
generate variations on a DNA string, and present pseudocode
for the recursive backtracking method that we use here.</p>
<p>In <a href="#">part 3 of this series</a>, we will cover our implementation
of the recursive backtracking method using two of Go's
unique features: channels, and Go routines. We implement
code that can utilize concurrency to generate variations
efficiently and collect results using a channel.</p>
<p><a name="perms-combs-vars"></a></p>
<h2>Permutations vs Combinations vs Variations</h2>
<p>Before covering generation of variations of a DNA string,
we should cover some terminology for clarification.</p>
<p>If we were to use the term <em>permutations</em>, as in, we are
counting (or generating) <em>permutations</em> of the input DNA string, 
that would imply that we were doing some kind of
rearrangement of the elements of the input DNA string
(for example, swapping two codons). This is not the
problem that we are solving, and requires different
formulas. (See <a href="#">Permutations</a>
entry on Wolfram MathWorld.)</p>
<p>The variations that we are referring to are not
exactly <em>combinations</em>, either, though. If we were
to use the term <em>combinations</em>, it would imply that
we were choosing a set of <span class="math">\(k\)</span> integers from a set
of <span class="math">\(d\)</span> integers <span class="math">\({1, 2, \dots, d}\)</span>.</p>
<p>The variations that we are counting are similar to
combinations, but with the additional act of swapping
out each codon at the position (integer) selected
with three other possible codons, so there are
more variations than combinations (and many
more permutations than variations).</p>
<p><a name="transform"></a></p>
<h2>Transforming the Problem Space</h2>
<p><a name="binary"></a></p>
<h3>Binary Numbers</h3>
<p><a name="list"></a></p>
<h2>List of Techniques</h2>
<p>Want to have an algorithm that will allow us to iterate over
every possible variation. This requires us to implement an
algorithm that can:</p>
<ul>
<li>Choose <span class="math">\(d\)</span> codons to modify from an input DNA string of length <span class="math">\(n\)</span></li>
<li>Swap out codons at these indices in every possible combination</li>
</ul>
<p>A general strategy is to reduce the problem to one
involving binary numbers, which can be implemented
in any decent programming language, are compact and 
simple, and translate the problem into a more universal
language of binary numbers.</p>
<p><a name="binary">&gt;</a></p>
<h3>Generating Binary Numbers with Constraints</h3>
<p>To generate the codons to modify,
want to generate numbers with <span class="math">\(n\)</span> digits
and <span class="math">\(d\)</span> bits (1s). We swap out the codons
at the positions with 1s.</p>
<p>Algorithm: <a href="https://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation">next bit permutation</a></p>
<p><a name="bitvectors"></a></p>
<h3>Bitvectors</h3>
<p>More a brief aside than anything else, as this is a very
large topic that we don't have the space to cover,
but we should mention here that bitvectors are 
the data structure of choice for dealing with
binary numbers and manipulating them for the
algorithms covered here.</p>
<p>For more information, see the following:</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Bit_array">Bitvector (wikipedia)</a></li>
<li><a href="https://graphics.stanford.edu/~seander/bithacks.html">Bit Twiddling Hacks</a></li>
</ul>
<p><a name="recursion"></a></p>
<h3>Recursion</h3>
<p>Recursion is a common pattern to use for problems that require
exploring a large problem space that requires us to make
several selections.</p>
<p>A recursive backtracking algorithm is analogous to exploring a
maze but laying out a rope as you go, so tht you can revisit
each possible route. In this case, we are using backtracking
to make the choice of which indices of the input DNA string
to modify. We want to explore all possible choices to generate
all possible variations of the input DNA string, and backtracking
gives us the framework to do that.</p>
<p><a name="backtracking"></a></p>
<h2>Recursive Backtracking Pseudocode</h2>
<p>Basic pseudocode for a backtracking method:</p>
<div class="highlight"><pre><span></span>explore method:
    base case:
        visit this solution
    recursive case:
        make a choice
        explore the consequences
        unmake the choice
        move on to the next choice
</pre></div>


<p><a name="dna"></a></p>
<h2>Applying to DNA Variations</h2>
<p>There are actually two places where we need to apply
backtracking to our problem.</p>
<ul>
<li><a href="#generating-visits">Generating Visits</a></li>
<li><a href="#assembling">Assembling the Variation</a></li>
</ul>
<script type="text/javascript">if (!document.getElementById('mathjaxscript_pelican_#%@#$@#')) {
    var align = "center",
        indent = "0em",
        linebreak = "false";

    if (false) {
        align = (screen.width < 768) ? "left" : align;
        indent = (screen.width < 768) ? "0em" : indent;
        linebreak = (screen.width < 768) ? 'true' : linebreak;
    }

    var mathjaxscript = document.createElement('script');
    mathjaxscript.id = 'mathjaxscript_pelican_#%@#$@#';
    mathjaxscript.type = 'text/javascript';
    mathjaxscript.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.3/latest.js?config=TeX-AMS-MML_HTMLorMML';
    mathjaxscript[(window.opera ? "innerHTML" : "text")] =
        "MathJax.Hub.Config({" +
        "    config: ['MMLorHTML.js']," +
        "    TeX: { extensions: ['AMSmath.js','AMSsymbols.js','noErrors.js','noUndefined.js'], equationNumbers: { autoNumber: 'AMS' } }," +
        "    jax: ['input/TeX','input/MathML','output/HTML-CSS']," +
        "    extensions: ['tex2jax.js','mml2jax.js','MathMenu.js','MathZoom.js']," +
        "    displayAlign: '"+ align +"'," +
        "    displayIndent: '"+ indent +"'," +
        "    showMathMenu: true," +
        "    messageStyle: 'normal'," +
        "    tex2jax: { " +
        "        inlineMath: [ ['\\\\(','\\\\)'] ], " +
        "        displayMath: [ ['$$','$$'] ]," +
        "        processEscapes: true," +
        "        preview: 'TeX'," +
        "    }, " +
        "    'HTML-CSS': { " +
        "        styles: { '.MathJax_Display, .MathJax .mo, .MathJax .mi, .MathJax .mn': {color: 'inherit ! important'} }," +
        "        linebreaks: { automatic: "+ linebreak +", width: '90% container' }," +
        "    }, " +
        "}); " +
        "if ('default' !== 'default') {" +
            "MathJax.Hub.Register.StartupHook('HTML-CSS Jax Ready',function () {" +
                "var VARIANT = MathJax.OutputJax['HTML-CSS'].FONTDATA.VARIANT;" +
                "VARIANT['normal'].fonts.unshift('MathJax_default');" +
                "VARIANT['bold'].fonts.unshift('MathJax_default-bold');" +
                "VARIANT['italic'].fonts.unshift('MathJax_default-italic');" +
                "VARIANT['-tex-mathit'].fonts.unshift('MathJax_default-italic');" +
            "});" +
            "MathJax.Hub.Register.StartupHook('SVG Jax Ready',function () {" +
                "var VARIANT = MathJax.OutputJax.SVG.FONTDATA.VARIANT;" +
                "VARIANT['normal'].fonts.unshift('MathJax_default');" +
                "VARIANT['bold'].fonts.unshift('MathJax_default-bold');" +
                "VARIANT['italic'].fonts.unshift('MathJax_default-italic');" +
                "VARIANT['-tex-mathit'].fonts.unshift('MathJax_default-italic');" +
            "});" +
        "}";
    (document.body || document.getElementsByTagName('head')[0]).appendChild(mathjaxscript);
}
</script></div>

    </div>
</div>

<div class="v20"></div>

<div class="row">
    <div class="col-md-12" style="text-align: left;">

        <div class="tags">
            <p>Tags:&nbsp;&nbsp;&nbsp;
                    <a href="/tag/go.html">go</a>
                    &nbsp;&nbsp;
                    <a href="/tag/golang.html">golang</a>
                    &nbsp;&nbsp;
                    <a href="/tag/rosalind.html">rosalind</a>
                    &nbsp;&nbsp;
                    <a href="/tag/bioinformatics.html">bioinformatics</a>
                    &nbsp;&nbsp;
                    <a href="/tag/recursion.html">recursion</a>
                    &nbsp;&nbsp;
                    <a href="/tag/backtracking.html">backtracking</a>
                    &nbsp;&nbsp;
                    <a href="/tag/strings.html">strings</a>
                    &nbsp;&nbsp;
                    <a href="/tag/combinatorics.html">combinatorics</a>
                    &nbsp;&nbsp;
        </div>

    </div>
</div>


<div class="v200"></div>

</div>


<script type="text/javascript" src="/theme/js/jquery-1.11.2.js"></script>
<script type="text/javascript" src="/theme/js/bootstrap-3.3.4.js"></script>
<script type="text/javascript" src="/theme/js/d3-3.5.5.js"></script>
<!--
<script type="text/javascript" src="/theme/js/leaflet-0.7.3.js"></script>
<script type="text/javascript" src="/theme/js/leaflet.ajax.min.js"></script>
--><script type="text/javascript" src="/theme/js/jquery-1.11.2.js"></script>
<script type="text/javascript" src="/theme/js/bootstrap-3.3.4.js"></script>

<footer id="contentinfo" class="body">
<p>&nbsp;</p>
<p>&nbsp;</p>
<center>
	<hr />
</center>
<p>&nbsp;</p>
<p style="text-align: center">
    <span class="fa-stack fa-lg">
        <i class="fa fa-square fa-stack-2x" style="color:#000;"></i>
        <i class="fa fa-terminal fa-stack-1x fa-inverse"></i>
    </span>
    Made from the command line with vim by 
    <a href="http://charlesreid1.com">charlesreid1</a><br />
    with help from <a href="https://getbootstrap.com/">Bootstrap</a> and <a href="http://getpelican.com">Pelican</a>.
</p>

<br />

<p style="text-align: center">
    <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">
    <span class="fa-stack fa-lg">
        <i class="fa fa-square fa-stack-2x" style="color:#000;"></i>
        <i class="fa fa-creative-commons fa-stack-1x fa-inverse"></i>
    </span>
    </a>
    <br />
    Licensed under the <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 License</a>.
</p>

<!-- Start of StatCounter Code for Default Guide -->
<script type="text/javascript">
var sc_project=11611748;
var sc_invisible=1;
var sc_security="9fcba820";
</script>
<script type="text/javascript"
src="https://www.statcounter.com/counter/counter.js"
async></script>

<noscript><div class="statcounter"><a title="Web Analytics"
href="http://statcounter.com/" target="_blank"><img
class="statcounter"
src="//c.statcounter.com/11611748/0/9fcba820/1/" alt="Web
Analytics"></a></div></noscript>
<!-- End of StatCounter Code for Default Guide -->
</footer><!-- /#contentinfo -->
</body>
</html>